src/posit/connect/_api.py

```diff
@@ -0,0 +1,278 @@
+
+from __future__ import annotations
+
+import itertools
+import posixpath
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, Generator, Generic, Optional, TypeVar, cast, overload
+
+from ._api_call import ApiCallMixin, get_api
+from ._json import Jsonifiable, JsonifiableDict, ResponseAttrs
+
+if TYPE_CHECKING:
+    from .context import Context
+
+
+# Design Notes:
+# * Perform API calls on property retrieval. e.g. `my_content.repository`
+# * Dictionary endpoints: Retrieve all attributes during init unless provided
+# * List endpoints: Do not retrieve until `.fetch()` is called directly. Avoids cache invalidation issues.
+#   * While slower, all ApiListEndpoint helper methods should `.fetch()` on demand.
+# * Only expose methods needed for `ReadOnlyDict`.
+#   * Ex: When inheriting from `dict`, we'd need to shut down `update`, `pop`, etc.
+# * Use `ApiContextProtocol` to ensure that the class has the necessary attributes for API calls.
+#    * Inherit from `ApiCallMixin` to add all helper methods for API calls.
+# * Classes should write the `path` only once within its init method.
+#    * Through regular interactions, the path should only be written once.
+
+# When making a new class,
+# * Use a class to define the parameters and their types
+#    * If attaching the type info class to the parent class, start with `_`. E.g.: `ContentItemRepository._Attrs`
+# * Document all attributes like normal
+#    * When the time comes that there are multiple attribute types, we can use overloads with full documentation and unpacking of type info class for each overload method.
+# * Inherit from `ApiDictEndpoint` or `ApiListEndpoint` as needed
+#    * Init signature should be `def __init__(self, ctx: Context, path: str, /, **attrs: Jsonifiable) -> None:`
+
+
+# This class should not have typing about the class keys as that would fix the class's typing. If
+# for some reason, we _know_ the keys are fixed (as we've moved on to a higher version), we can add
+# `Generic[AttrsT]` to the class.
+class ReadOnlyDict(Mapping):
+    _attrs: ResponseAttrs
+    """Resource attributes passed."""
+
+    def __init__(self, attrs: ResponseAttrs) -> None:
+        """
+        A read-only dict abstraction for any HTTP endpoint that returns a singular resource.
+
+        Parameters
+        ----------
+        attrs : dict
+            Resource attributes passed
+        """
+        super().__init__()
+        self._attrs = attrs
+
+    def get(self, key: str, default: Any = None) -> Any:
+        return self._attrs.get(key, default)
+
+    def __getitem__(self, key: str) -> Any:
+        return self._attrs[key]
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        raise AttributeError(
+            "Resource attributes are locked. "
+            "To retrieve updated values, please retrieve the parent object again."
+        )
```

Comment:
I believe this should be a `NotImplementedError` instead of an `AttributeError`.

Reply:
👍